File: backend/app/schemas/applications.py
from typing import Literal

from pydantic import BaseModel

from app.ai.ai_services.ats_service import (
    ATSCategoryScore as _ATSCategoryScore,
    ATSWeightage as _ATSWeightage,
    GraceCredit as ATSGraceCredit,
    RelevantExperienceResult as ATSRelevantExperience,
    RequirementMatchResult as _RequirementMatchResult,
    ResponsibilityMatchResult as ATSResponsibilityMatchItem,
    SectionMatchResult as ATSSectionMatchItem,
)


class ApplyRequest(BaseModel):
    job_posting_id: str
    candidate_id: str


# ---------------------------------------------------------------------------
# ATS result — the field shapes below are owned by app.ai.ai_services.ats_service
# (the AI service that produces a fresh ATS result); these subclasses only relax
# a few fields to Optional/defaulted so _parse_ats_details() can degrade legacy or
# malformed stored `ats_details` JSON to a partial/None result instead of raising,
# without needing a second, hand-maintained copy of every field.
# ---------------------------------------------------------------------------


class ATSCategoryScore(_ATSCategoryScore):
    included: bool = True


class ATSWeightage(_ATSWeightage):
    experience: ATSCategoryScore
    skills: ATSCategoryScore
    projects: ATSCategoryScore
    certifications: ATSCategoryScore
    education: ATSCategoryScore
    achievements: ATSCategoryScore
    qualify_threshold: float = 65
    pass_fail: Literal["PASS", "FAIL"] | None = None


class ATSRequirementMatchItem(_RequirementMatchResult):
    category: Literal["skills", "projects", "certifications", "education", "achievements"] = "skills"
    max_score: float


class ATSCheckResponse(BaseModel):
    verdict: Literal["QUALIFIED", "UNDERQUALIFIED", "OVERQUALIFIED"]
    verdict_summary: str
    weightage: ATSWeightage | None = None
    requirement_matching: list[ATSRequirementMatchItem] = []
    relevant_experience: ATSRelevantExperience | None = None
    section_matching: list[ATSSectionMatchItem] = []
    grace_credits: list[ATSGraceCredit] = []
    additional_cv_content: list[str] = []
    is_overqualified: bool = False
    final_verdict: Literal["PASS", "FAIL"] | None = None
    override_reason: str | None = None


class ApplyResponse(BaseModel):
    application_id: str
    created: bool
    message: str


class InterviewRoundInfo(BaseModel):
    interview_round_id: str
    interview_id: str | None
    title: str
    round_order: int
    status: str | None
    feedback: str | None = None
    result: float | None = None
    scheduled_at: str | None = None
    completed_at: str | None = None
    avg_score: float | None = None


class InterviewStagesResponse(BaseModel):
    application_id: str
    rounds: list[InterviewRoundInfo]
    current_round_id: str | None  # interview_round_id of the active round
    ats_status: str  # 'pending' | 'pass' | 'fail'
    ats_result: ATSCheckResponse | None = None
    application_status: str | None = None  # raw DB value: ATS_PENDING/ATS_PASS/ATS_FAIL/IN_PROGRESS/HIRED/REJECTED
    job_role_title: str | None = None
    experience_level_name: str | None = None
    company: str | None = None


class InterviewQuestionItem(BaseModel):
    question_id: str
    question_text: str
    candidate_answer: str | None = None
    score: int | None = None
    notes: str | None = None


class MyApplicationItem(BaseModel):
    application_id: str
    job_id: str
    status: str
    applied_at: str
    description: str
    location: str | None
    job_type: str
    salary_range: str | None
    job_role_id: int
    job_role_title: str
    experience_level_id: int
    experience_level_name: str
    company: str
    latest_interview_round_title: str | None = None

File: frontend/src/api/applications.ts
const BASE_URL = (import.meta.env.VITE_API_URL as string | undefined) ?? 'http://localhost:8000'

export interface ATSCategoryScore {
  score_pct: number
  weight_pct: number
  weighted_score: number
  included: boolean
  matched_count: number | null
  total_count: number | null
}

export interface ATSWeightage {
  experience: ATSCategoryScore
  skills: ATSCategoryScore
  projects: ATSCategoryScore
  certifications: ATSCategoryScore
  education: ATSCategoryScore
  achievements: ATSCategoryScore
  weighted_average: number
  qualify_threshold: number
  pass_fail: 'PASS' | 'FAIL' | null
}

export type ATSMatchType = 'exact' | 'parent' | 'alternative' | 'exceeds' | 'not_found' | 'not_required'
export type ATSConfidence = 'high' | 'medium' | 'low'
export type ATSRequirementCategory = 'skills' | 'projects' | 'certifications' | 'education' | 'achievements'

export interface ATSRequirementMatchItem {
  requirement: string
  category: ATSRequirementCategory
  candidate_evidence: string | null
  match_type: ATSMatchType
  exceeds_requirement: boolean
  score: number
  max_score: number
  confidence: ATSConfidence
  reason: string
}

export interface ATSResponsibilityMatchItem {
  requirement: string
  evidence: string | null
  match_type: 'direct' | 'close' | 'not_found'
  score: number
}

export interface ATSRelevantExperience {
  job_role_required: string
  included_experience: string
  excluded_experience: string | null
  required_years: number | null
  candidate_relevant_years: number | null
  status: 'qualified' | 'underqualified' | 'overqualified'
  responsibility_matches: ATSResponsibilityMatchItem[]
}

export interface ATSSectionMatchItem {
  section: string
  jd_requires_section: boolean
  cv_has_content: boolean
  score: number
  note: string
}

export interface ATSGraceCredit {
  item: string
  points: number
  note: string
}

export interface ATSCheckResponse {
  verdict: 'QUALIFIED' | 'UNDERQUALIFIED' | 'OVERQUALIFIED'
  verdict_summary: string
  weightage: ATSWeightage | null
  requirement_matching: ATSRequirementMatchItem[]
  relevant_experience: ATSRelevantExperience | null
  section_matching: ATSSectionMatchItem[]
  grace_credits: ATSGraceCredit[]
  additional_cv_content: string[]
  is_overqualified: boolean
  final_verdict: 'PASS' | 'FAIL' | null
  override_reason: string | null
}

export interface ATSRerunNotice {
  previous_verdict: string | null
  new_verdict: string
  message: string
}

export interface MyApplicationItem {
  application_id: string
  job_id: string
  status: string
  applied_at: string
  description: string
  location: string | null
  job_type: string
  salary_range: string | null
  job_role_id: number
  job_role_title: string
  experience_level_id: number
  experience_level_name: string
  company: string
  latest_interview_round_title: string | null
}

export async function runAts(applicationId: string): Promise<ATSCheckResponse> {
  const res = await fetch(`${BASE_URL}/api/applications/${applicationId}/run-ats`, {
    method: 'POST',
  })
  if (!res.ok) throw new Error('ATS check failed')
  return res.json() as Promise<ATSCheckResponse>
}

export async function fetchMyApplications(candidateId: string): Promise<MyApplicationItem[]> {
  const res = await fetch(
    `${BASE_URL}/api/applications/my-applications?candidate_id=${encodeURIComponent(candidateId)}`,
  )
  if (!res.ok) throw new Error('Failed to load your applications')
  return res.json() as Promise<MyApplicationItem[]>
}

export async function ackAtsRerunNotice(applicationId: string): Promise<void> {
  await fetch(`${BASE_URL}/api/applications/${encodeURIComponent(applicationId)}/ack-ats-rerun`, {
    method: 'POST',
  })
}
